x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/SqlQueryResponse.java

```diff
@@ -81,10 +82,16 @@ public SqlQueryResponse(StreamInput in) throws IOException {
             }
         }
         this.rows = unmodifiableList(rows);
-        columnar = in.readBoolean();
-        asyncExecutionId = in.readOptionalString();
-        isPartial = in.readBoolean();
-        isRunning = in.readBoolean();
+        if (in.getVersion().onOrAfter(Version.V_7_14_0)) {
+            columnar = in.readBoolean();
+            asyncExecutionId = in.readOptionalString();
+            isPartial = in.readBoolean();
+            isRunning = in.readBoolean();
+        } else {
+            asyncExecutionId = null;
+            isPartial = false;
+            isRunning = false;
+        }
```

Comment:
Why this change and why is `columnar` included there (it seems you grouped together the async specific settings)?

Reply:
It hadn't been serialised before the given version: https://github.com/elastic/elasticsearch/pull/73991/files#diff-3c77d10764f4a72e650dfabe508e5d372bf024a776ca17b81c2f4666f859fe01R187 , which was probably an oversight.

Reply:
The fact that we discovered this in a BWC test and not when the async search was introduced is troubling. At some point in future we will need a set of tests that should cover the serialization differences between bwc versions for both the request and response objects. I did something similar for EQL here: https://github.com/elastic/elasticsearch/pull/68339. @Luegg please open an issue for improving the tests coverage in this area.

Reply:
My initial question still stands.
To clarify a bit: `columnar` has been added in 7.2.0 (https://github.com/elastic/elasticsearch/pull/39287). If we should handle it differently depending on the version it shouldn't be 7.14, since that's not the version that added it. Or I am missing something.

Reply:
> it shouldn't be 7.14, since that's not the version that added it.

True, but that's the version that added it to the serialisation.
If not mistaking, between (and before) 7.2 and 7.14, a serialised `SqlQueryResponse` didn't contain the `columnar` field. So trying to deserialise a `SqlQueryResponse` and read the `columnar` field would eventually fail.
A 7.14+ serialised `SqlQueryResponse` will otoh contain it and needs to be read.

Reply:
Thanks @bpintea, I was getting confused myself.